File: numbers/src/test/java/mathtools/numbers/primes/cache/ShortPrimeCacheTest.java
package mathtools.numbers.primes.cache;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.google.common.collect.Lists;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.RepeatedTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.util.List;

import kotlin.random.Random;
import mathtools.numbers.primes.BytePrimeCache;
import mathtools.numbers.primes.PrimeTestDataProvider;
import mathtools.numbers.primes.ShortPrimeCache;
import mathtools.pairs.IntPair;

/** Testing [ShortPrimeCache] class
 * @author DK96-OS : 2022 */
public final class ShortPrimeCacheTest {

	private final List<Integer> primeList = PrimeTestDataProvider.getPrimes509();

	private final int maxPrimeIndex = primeList.size() - 1;

	private ShortPrimeCache cache = null;

	@BeforeEach
	public void testSetup() {
		cache = new ShortPrimeCache();
	}

	@Test
	public void testGetPrimeDecreasing() {
		for (
			int i = maxPrimeIndex;
			0 <= i;
			--i
		) {
			assertEquals(
				primeList.get(i),
				cache.getPrime(i)
			);
			cache.clear();
		}
	}

	@Test
	public void testSpecificPrimes() {
		assertEquals(
			547, cache.getPrime(100));
		assertEquals(
			661, cache.getPrime(120));
		assertEquals(
			811, cache.getPrime(140));
		assertEquals(
			947, cache.getPrime(160));
		assertEquals(
			1087, cache.getPrime(180));
		assertEquals(
			1229, cache.getPrime(200));
		assertEquals(
			1993, cache.getPrime(300));
		assertEquals(
			2131, cache.getPrime(320));
		assertEquals(
			2293, cache.getPrime(340));
		assertEquals(
			2437, cache.getPrime(360));
		assertEquals(
			2621, cache.getPrime(380));
		assertEquals(
			2749, cache.getPrime(400));
		assertEquals(
			5801, cache.getPrime(760));
		assertEquals(
			7001, cache.getPrime(900));
		assertEquals(
			7211, cache.getPrime(920));
		assertEquals(
			7727, cache.getPrime(980));
	}

	@Test
	public void testIsPrime() {
		int n = 2;
		for (;
		     n < primeList.get(maxPrimeIndex);
		     ++n
		) assertEquals(
			primeList.contains(n),
			cache.isPrime(n)
		);
	}

	@Test
	public void testIsPrimeCacheClear() {
		for (
			int n = primeList.get(maxPrimeIndex);
			2 <= n;
			--n
		) {
			cache.clear();
			assertEquals(
				primeList.contains(n),
				cache.isPrime(n)
			);
		}
	}

	@Test
	public void testLargeIndices() {
		final List<IntPair> testData = Lists.newArrayList(
			new IntPair(1000, 7927),
			new IntPair(2000, 17393),
			new IntPair(3000, 27457),
			new IntPair(3511, 32749)
		);
		for (final IntPair pair : testData) {
			assertEquals(
				pair.getSecond(),
				cache.getPrime(pair.getFirst())
			);
		}
	}

	@Test
	public void testPrimeTooLargeForCache() {
		assertThrows(
			IllegalStateException.class,
			() -> cache.getPrime(cache.getMaxIndex() + 5)
		);
	}

	@ParameterizedTest
	@ValueSource(ints = {
		1, 2, 3, 5, 6, 8, 10, 15
	})
	public void testExpansionRates(
		final int rate
	) {
		for (
			int i = 0;
			maxPrimeIndex >= i;
			i += rate
		) assertEquals(
			primeList.get(i),
			cache.getPrime(i)
		);
	}

	@RepeatedTest(2)
	public void testRandomAccess() {
		for (
			int i = 0;
			200 > i;
			++i
		) {
			final int rand = Random.Default.nextInt(
				1, 70
			);
			assertEquals(
				primeList.get(rand),
				cache.getPrime(rand)
			);
			cache.clear();
		}
	}

	@RepeatedTest(2)
	public void testRandomIsPrime() {
		final int maxIndex = cache.getMaxIndex();
		for (
			int x = 0;
			20 > x;
			++x
		) {
			cache.clear();
			for (
				int i = 0;
				100 > i;
				++i
			) {
				final int randIndex = Random.Default.nextInt(
					BytePrimeCache.MAX_INDEX + 1,
					maxIndex
				);
				final int rand = cache.getPrime(randIndex);
				assertTrue(
					cache.isPrime(rand));
			}
		}
	}

}

File: numbers/src/test/kotlin/mathtools/numbers/primes/ShortPrimeCacheTest.kt
package mathtools.numbers.primes

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.RepeatedTest
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

/** Testing the Short PrimeCache class, for primes up to 32767 */
class ShortPrimeCacheTest {
	
	private val primeList: List<Int> = PrimeTestDataProvider.getPrimes349()
    /* These are the indices for a selection of numbers in the prime list:
    	assertEquals(199, primeList[45])
    	assertEquals(251, primeList[53])
    	assertEquals(349, primeList[69])
    */

	@Test
	fun testIsPrime() {
		val cache = ShortPrimeCache()
		for (n in 2 until primeList.last())
			assertEquals(n in primeList, cache.isPrime(n))
		for (n in primeList.last() downTo 2) {
			cache.clear()
			assertEquals(n in primeList, cache.isPrime(n))
		}
	}
	
	@Test
	fun testGetPrimeDecreasing() {
		val cache = ShortPrimeCache()
		for (i in primeList.size - 1 downTo 0) {
			assertEquals(primeList[i], cache.getPrime(i))
			cache.clear()
		}
	}

	@ParameterizedTest
	@ValueSource(ints = [1, 2, 3, 5, 6, 8, 10, 15])
	fun testExpansionRates(rate: Int) {
		val cache = ShortPrimeCache()
		for (i in 0 until primeList.size step rate)
			assertEquals(primeList[i], cache.getPrime(i))
	}

	@Test
	fun testSpecificPrimes() {
		val cache = ShortPrimeCache()
		assertEquals(547, cache.getPrime(100))
		assertEquals(661, cache.getPrime(120))
		assertEquals(811, cache.getPrime(140))
		assertEquals(947, cache.getPrime(160))
		assertEquals(1087, cache.getPrime(180))
		assertEquals(1229, cache.getPrime(200))
		assertEquals(1993, cache.getPrime(300))
		assertEquals(2131, cache.getPrime(320))
		assertEquals(2293, cache.getPrime(340))
		assertEquals(2437, cache.getPrime(360))
		assertEquals(2621, cache.getPrime(380))
		assertEquals(2749, cache.getPrime(400))
		assertEquals(5801, cache.getPrime(760))
		assertEquals(7001, cache.getPrime(900))
		assertEquals(7211, cache.getPrime(920))
		assertEquals(7727, cache.getPrime(980))
	}

	@Test
	fun testShort() {
		println("Short: ${Short.MAX_VALUE}")
	}

	@Test
	fun testMaxIndex() {
		val cache = ShortPrimeCache()
		val testIndices = mutableListOf(
			1000, 2000, 3000, 3090, 3490,
			3510, 3511
		)
		for (i in testIndices) {
			val prime = cache.getPrime(i)
			println("Query($i) => $prime")
		}
	}

/*
	@Test fun testCacheExpansionSlow() {
		val cache = ShortPrimeCache()
		assertEquals(137, cache.getPrime(32))
		assertEquals(139, cache.getPrime(33))
		assertEquals(149, cache.getPrime(34))
		assertEquals(3, cache.arraySize)
		assertEquals(1, cache.queueSize)
		assertEquals(151, cache.getPrime(35))
		assertEquals(2, cache.queueSize)
		assertEquals(157, cache.getPrime(36))
		assertEquals(3, cache.queueSize)
		for (i in 37 .. 50)
			assertEquals(primeList[i], cache.getPrime(i))
	}
	@Test fun testCacheExpansionQuick() {
		val cache = ShortPrimeCache()
		assertEquals(3, cache.arraySize)
		assertEquals(0, cache.queueSize)
			// Expand the cache by 3
		assertEquals(157, cache.getPrime(36))
		assertEquals(3, cache.arraySize)
		assertEquals(3, cache.queueSize)
		assertEquals(149, cache.getPrime(34))
		assertEquals(139, cache.getPrime(33))
		assertEquals(151, cache.getPrime(35))
		for (i in 36 until 58 step 8)
			assertEquals(primeList[i], cache.getPrime(i))
	}
*/
	
	@RepeatedTest(4) 
	fun testRandomAccess() {
		val cache = ShortPrimeCache()
		val testIndexRange = 1 .. 69
		for (i in 0 until 200) {
			val rand = testIndexRange.random()
			assertEquals(primeList[rand], cache.getPrime(rand))
			cache.clear()
		}
		for (x in 0 until 20) {
			cache.clear()
			for (i in 0 until 100) {
				val rand = cache.getPrime(cache.shortIndexRange.random())
				assertEquals(true, cache.isPrime(rand))
			}
		}
	}

	@Test
	fun testPrimeTooLargeForCache() {
		val cache = ShortPrimeCache()
		assertThrows<IllegalStateException> {
			cache.getPrime(cache.maxIndex + 5)
		}
	}

}